packages/flutter/lib/src/gestures/mouse_tracking.dart

```diff
@@ -219,75 +221,90 @@ class MouseTracker extends ChangeNotifier {
+            exitAnnotation(trackedAnnotation, deviceId);
           }
-        }
-        if (hitAnnotation.annotation?.onHover != null && lastEvent is PointerHoverEvent) {
-          hitAnnotation.annotation.onHover(lastEvent);
+          continue;
         }
 
-        // Tell any tracked annotations that weren't hit that they are no longer
-        // active.
-        for (_TrackedAnnotation trackedAnnotation in _trackedAnnotations.values) {
-          if (hitAnnotations.contains(trackedAnnotation)) {
-            continue;
+        final Set<_TrackedAnnotation> hitAnnotations = hits.map<_TrackedAnnotation>((MouseTrackerAnnotation hit) => _findAnnotation(hit)).toSet();
+        for (_TrackedAnnotation hitAnnotation in hitAnnotations) {
+          if (!hitAnnotation.activeDevices.contains(deviceId)) {
+            // A tracked annotation that just became active and needs to have an enter
+            // event sent to it.
+            hitAnnotation.activeDevices.add(deviceId);
+            if (hitAnnotation.annotation?.onEnter != null) {
+              hitAnnotation.annotation.onEnter(PointerEnterEvent.fromMouseEvent(lastEvent));
+            }
           }
-          if (trackedAnnotation.activeDevices.contains(deviceId)) {
-            if (trackedAnnotation.annotation?.onExit != null) {
-              trackedAnnotation.annotation.onExit(PointerExitEvent.fromMouseEvent(lastEvent));
+          if (hitAnnotation.annotation?.onHover != null && lastEvent is PointerHoverEvent) {
+            hitAnnotation.annotation.onHover(lastEvent);
+          }
+
+          // Tell any tracked annotations that weren't hit that they are no longer
+          // active.
+          for (_TrackedAnnotation trackedAnnotation in _trackedAnnotations.values) {
+            if (hitAnnotations.contains(trackedAnnotation)) {
+              continue;
+            }
+            if (trackedAnnotation.activeDevices.contains(deviceId)) {
+              if (trackedAnnotation.annotation?.onExit != null) {
+                trackedAnnotation.annotation.onExit(PointerExitEvent.fromMouseEvent(lastEvent));
+              }
+              trackedAnnotation.activeDevices.remove(deviceId);
             }
-            trackedAnnotation.activeDevices.remove(deviceId);
           }
         }
       }
+    } finally {
+      _pendingRemovals.clear();
     }
   }
 
   void _addMouseEvent(int deviceId, PointerEvent event) {
     final bool wasConnected = mouseIsConnected;
+    if (event is PointerAddedEvent) {
+      // If we are adding the device again, then we're not removing it anymore.
+      _pendingRemovals.remove(deviceId);
+    }
     _lastMouseEvent[deviceId] = event;
     if (mouseIsConnected != wasConnected) {
       notifyListeners();
     }
   }
 
-  void _removeMouseEvent(int deviceId) {
+  void _removeMouseEvent(int deviceId, PointerEvent event) {
     final bool wasConnected = mouseIsConnected;
+    if (event is PointerRemovedEvent) {
```

Comment:
Wondering if there's ever a way to remove a mouse event that's not a PointerRemovedEvent? If there isn't maybe this could just be an assert and we always add things to _pendingRemovals?

Reply:
Converted to assert.